### PR TITLE
Implement Automatic Exception Reporting (AER) - READ CAREFULLY

### PR DESCRIPTION
**Pull request by bubby932**
* Bumps firebase-admin.
* Implements /api/analytics/exception-report

This pull request CANNOT be merged until #411  is resolved, otherwise we will be in direct violation of our privacy policy, risking legal action. 

Notes about the AER system:
* AER is entirely opt-in, and disabled by default.
* Clientside and serverside protections are both in place to prevent users from being able to send exception reports in violation of their privacy settings.
* All exception reports are encrypted with RSA ([see package](https://www.npmjs.com/package/node-rsa)) with a public key *only*. Private keys are stored *only* on a server which has no inbound HTTP access to the wider internet, and is accessable only by highly privileged team members.
* Users may opt-in to allow our team to contact them (which adds their user ID to each exception log), however, having this feature **does not** guarantee flawless redaction of PII, HWID, or other sensitive information. Therefore, the encryption and high security dedicated to this system remains necessary.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -5,4 +5,5 @@ src
 env.js
 nginx.conf
 data/audit.json
-admin.json
\ No newline at end of file
+admin.json
+keys/
```
Comment (bubby932) [inline .gitignore:9]: While not entirely necessary, it is helpful to prevent leaking the public key in any circumstance.

File: package.json [hunk 2]
```diff
@@ -21,11 +21,12 @@
     "express-rate-limit": "^6.6.0",
     "express-slow-down": "^1.5.0",
     "firebase": "^9.12.1",
-    "firebase-admin": "^11.1.0",
+    "firebase-admin": "^11.2.0",
```
Comment (bubby932) [inline package.json:24]: Bumped by @dependabot.

File: package.json [hunk 3]
```diff
@@ -21,11 +21,12 @@
...
     "fuse.js": "^6.5.3",
     "jsonwebtoken": "^8.5.1",
     "mongodb": "^4.10.0",
     "node-cache": "^5.1.2",
+    "node-rsa": "^1.1.1",
```
Comment (bubby932) [inline package.json:29]: Used for RSA encryption of exception reports.

File: routers/analytics.js [hunk 4]
```diff
@@ -1,6 +1,12 @@
 const router = require('express').Router();
+const { PullPlayerData } = require('../helpers');
+const { authenticateToken } = require('../middleware');
 const { GetInstances } = require('./matchmaking');
+const { readFileSync } = require('node:fs');
+const RSA = require('node-rsa');
+const cfg = require('../config.json');
 
+const EXCEPTION_LOGGING_PUBLIC_KEY = new RSA().importKey(readFileSync(cfg.exception_logging_publickey_path).toString('utf-8'), cfg.exception_logging_publickey_format);
```
Comment (bubby932) [inline routers/analytics.js:9]: As you can see, the only encryption key even vaguely accessible from the greater internet is the public key, which should be insignificant.

File: routers/analytics.js [hunk 5]
```diff
@@ -19,4 +25,50 @@ router.get("/instance-count", async (req, res) => {
     res.status(200).send(`${instances.length}`);
 });
 
+router.put("/exception-report", authenticateToken, async (req, res) => {
```
Comment (bubby932) [inline routers/analytics.js:28]: AER primary endpoint.

File: routers/analytics.js [hunk 6]
```diff
@@ -19,4 +25,50 @@ router.get("/instance-count", async (req, res) => {
...
+    let data = await PullPlayerData(req.user.id);
+    if (data.settings.ALLOW_EXCEPTION_REPORTING !== "PERMIT") {
```
Comment (bubby932) [inline routers/analytics.js:30]: Double-check user settings before proceeding, once client-side, once server-side.

File: routers/analytics.js [hunk 7]
```diff
@@ -19,4 +25,50 @@ router.get("/instance-count", async (req, res) => {
...
+        throw new Error("Exception report attempted when user does not permit automatic exception reporting - this is a privacy violation.");
+    }
+
+    try {
+        const exception_data = {
+            message: req.body.message,
+            stack_trace: req.body.stack_trace
+        };
+
+        if (data.settings.EXCEPTION_REPORTING_DONT_ANONYMIZE === "true") {
```
Comment (bubby932) [inline routers/analytics.js:47]: "Allow CVR team to contact me about automatic reports" setting.

File: routers/analytics.js [hunk 8]
```diff
@@ -19,4 +25,50 @@ router.get("/instance-count", async (req, res) => {
...
+            exception_data.PLAYER_ID = req.user.id;
+        }
+
+        const encrypted = EXCEPTION_LOGGING_PUBLIC_KEY.encrypt(JSON.stringify(exception_data), 'buffer');
```
Comment (bubby932) [inline routers/analytics.js:51]: Encrypt data before uploading to database.

File: routers/analytics.js [hunk 9]
```diff
@@ -19,4 +25,50 @@ router.get("/instance-count", async (req, res) => {
...
+            .collection('exception_reports')
+            .insertOne({
+                data: encrypted
+            });
+        
+        return res.status(200).json({
+            code: "success",
+            message: "Operation successful. Thank you for the report."
+        });
+    } catch (ex) {
```
Comment (bubby932) [inline routers/analytics.js:65]: Critical error handling.

**Review by bubby932 — COMMENTED** [review]
Explained changes.

**Comment by bubby932** [comment]
#411 closed as complete, this PR can now proceed with the review & deployment process.

**Review by Bobrobot1 — APPROVED** [review]
Looks good.

**Pull request merged**